app/src/main/java/org/openobservatory/ooniprobe/domain/QueryDataSource.java

```diff
@@ -0,0 +1,63 @@
+package org.openobservatory.ooniprobe.domain;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.paging.PagingSource;
+import androidx.paging.PagingState;
+
+import com.raizlabs.android.dbflow.sql.language.SQLOperator;
+import com.raizlabs.android.dbflow.sql.language.SQLite;
+
+import org.openobservatory.ooniprobe.model.database.Result;
+import org.openobservatory.ooniprobe.model.database.Result_Table;
+
+import java.util.List;
+
+import kotlin.coroutines.Continuation;
+
+public class QueryDataSource extends PagingSource<Integer, Result> {
+    private final String testGroupNameFilter;
+
+    public QueryDataSource(String testGroupNameFilter) {
+        this.testGroupNameFilter = testGroupNameFilter;
+    }
+
+    @Nullable
+    @Override
+    public Integer getRefreshKey(@NonNull PagingState<Integer, Result> pagingState) {
+        return 1;
+    }
+
+    @Nullable
+    @Override
+    public LoadResult.Page<Integer, Result> load(@NonNull LoadParams<Integer> loadParams, @NonNull Continuation<? super LoadResult<Integer, Result>> continuation) {
+        // Key may be null during a refresh, if no explicit key is passed into Pager
+        // construction. Use 0 as default, because our API is indexed started at index 0
+        int pageNumber = loadParams.getKey() != null ? loadParams.getKey() : 0;
+
+        SQLOperator[] conditions = (testGroupNameFilter != null && !testGroupNameFilter.isEmpty())
+                ? new SQLOperator[]{Result_Table.test_group_name.is(testGroupNameFilter)}
+                : new SQLOperator[0];
+
+        List<Result> response = SQLite.select().from(Result.class)
+                .where(conditions)
+                .limit(20)
+                .offset(pageNumber * 20)
+                .orderBy(Result_Table.start_time, false)
+                .queryList();
```

Comment:
If while you have fetched the first page of data, another result is written to the results table (because there is a test currently running), fetching the next page will be off by one.

You could instead use the `result_id` as a pagination handler and query `WHERE result_id < $last_result_id` to get the next page worth of data.

Reply:
The query mechanism has been updated to use `result_id` based pagination